File: src/main/java/org/oewntk/model/Finder.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.model;

import java.util.*;

public class Finder
{
	/**
	 * Find lex matching word
	 *
	 * @param model model
	 * @param lemma lemma (CS)
	 * @return lexes
	 */
	public static Collection<Lex> getLexes(final CoreModel model, final String lemma)
	{
		return model.getLexesByLemma().get(lemma);
	}

	/**
	 * Find lex matching word and type filter
	 *
	 * @param model      model
	 * @param lemma      lemma (CS)
	 * @param typeFilter type filter
	 * @return lexes
	 */
	public static Lex[] getLexesHavingType(final CoreModel model, final String lemma, final char typeFilter)
	{
		return model.getLexesByLemma().get(lemma) //
				.stream().filter(lex -> lex.getType() == typeFilter) //
				.toArray(Lex[]::new);
	}

	/**
	 * Find lex matching word and part-of-speech filter
	 *
	 * @param model     model
	 * @param lemma     lemma (cased)
	 * @param posFilter part-of-speech filter
	 * @return lexes
	 */
	public static Lex[] getLexesHavingPos(final CoreModel model, final String lemma, final char posFilter)
	{
		return model.getLexesByLemma().get(lemma) //
				.stream().filter(lex -> lex.getPartOfSpeech() == posFilter) //
				.toArray(Lex[]::new);
	}

	/**
	 * Find lexes matching lemma ignoring case and having the desired type
	 *
	 * @param model      model
	 * @param lcLemma    lower-cased lemma
	 * @param typeFilter type filter
	 * @return array of lexes
	 */
	public static Lex[] getLcLexesHavingType(final CoreModel model, final String lcLemma, final Character typeFilter)
	{
		return model.getLexesByLCLemma().get(lcLemma.toLowerCase(Locale.ENGLISH)).stream() //
				.filter(lex -> lex.getType() == typeFilter) //
				.toArray(Lex[]::new);
	}

	/**
	 * Find lexes matching lemma ignoring case and having the desired pos
	 *
	 * @param model      model
	 * @param lcLemma      lower-cased lemma
	 * @param posFilter    pos filter
	 * @return array of lexes
	 */
	public static Lex[] getLcLexesHavingPos(final CoreModel model, final String lcLemma, final Character posFilter)
	{
		return model.getLexesByLCLemma().get(lcLemma.toLowerCase(Locale.ENGLISH)).stream() //
				.filter(lex -> lex.getPartOfSpeech() == posFilter) //
				.toArray(Lex[]::new);
	}

	/**
	 * Find lexes matching lemma ignoring case
	 *
	 * @param model      model
	 * @param lcLemma      lower-cased lemma
	 * @return array of lexes
	 */
	public static Lex[] getLcLexes(final CoreModel model, final String lcLemma)
	{
		return model.getLexesByLCLemma().get(lcLemma.toLowerCase(Locale.ENGLISH))
				.toArray(Lex[]::new);
	}

	/**
	 * Find lex matching pronunciations
	 *
	 * @param lexes          lexes
	 * @param pronunciations pronunciations
	 * @return lex
	 * @throws IllegalArgumentException if not found
	 */
	public static Lex getLexHavingPronunciations(Lex[] lexes, final Pronunciation[] pronunciations)
	{
		Set<Pronunciation> set = Utils.toSet(pronunciations);
		return Arrays.stream(lexes) //
				.filter(lex -> compareAsSets(set, lex.getPronunciations())) //
				.findFirst().orElseThrow(IllegalArgumentException::new);
	}

	/**
	 * Find lexes matching pronunciations
	 *
	 * @param lexes          lexes
	 * @param pronunciations pronunciations
	 * @return lex
	 * @throws IllegalArgumentException if not found
	 */
	public static Lex[] getLexesHavingPronunciations(final Lex[] lexes, final Pronunciation[] pronunciations)
	{
		Set<Pronunciation> set = Utils.toSet(pronunciations);
		return Arrays.stream(lexes) //
				.filter(lex -> compareAsSets(set, lex.getPronunciations())) //
				.toArray(Lex[]::new);
	}

	/**
	 * Find lex matching discriminant
	 *
	 * @param lexes        lexes
	 * @param discriminant discriminant
	 * @return lex
	 * @throws IllegalArgumentException if not found
	 */
	public static Lex getLexHavingDiscriminant(final Lex[] lexes, final String discriminant)
	{
		return Arrays.stream(lexes) //
				.filter(lex -> Objects.equals(lex.getDiscriminant(), discriminant)) //
				.findFirst().orElseThrow(IllegalArgumentException::new);
	}

	// Set comparison

	public static <T> boolean compareAsSets(T[] array1, T[] array2)
	{
		Set<T> set1 = Utils.toSet(array1);
		return compareAsSets(set1, array2);
	}

	public static <T> boolean compareAsSets(Set<T> set1, T[] array2)
	{
		Set<T> set2 = Utils.toSet(array2);
		return Objects.equals(set1, set2);
	}
}

File: src/main/java/org/oewntk/model/Finder.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

/**
 * Lex finder
 */
object Finder {

    /**
     * Find lex matching word
     *
     * @param model model
     * @param lemma lemma (CS)
     * @return collection of lexes
     */
    fun getLexes(model: CoreModel, lemma: Lemma): Collection<Lex> {
        return model.lexResolver(lemma)
    }

    /**
     * Find lex matching word and type filter
     *
     * @param model             model
     * @param word              word
     * @param targetCategory    target category
     * @param wordExtractor     word extractor
     * @param categoryExtractor category extractor
     * @return sequence of lexes
     */
    fun getLexesHaving(
        model: CoreModel,
        word: Lemma,
        targetCategory: Category,
        wordExtractor: (Lex) -> Lemma,
        categoryExtractor: (Lex) -> Category,
    ): Sequence<Lex> {
        return model.lexes
            .asSequence()
            .filter { word == wordExtractor.invoke(it) }
            .filter { targetCategory == categoryExtractor.invoke(it) }
    }

    /**
     * Find lex matching word and type filter
     *
     * @param model      model
     * @param lemma      lemma (CS)
     * @param targetType target type
     * @return sequence of lexes
     */
    fun getLexesHavingType(model: CoreModel, lemma: Lemma, targetType: SynsetType): Sequence<Lex>? {
        return model.lexFinder(lemma)
            ?.asSequence()
            ?.filter { it.type.value == targetType.value }
    }

    /**
     * Find lex matching word and part-of-speech filter
     *
     * @param model     model
     * @param lemma     lemma (cased)
     * @param targetPos target part-of-speech
     * @return sequence of lexes
     */
    fun getLexesHavingPos(model: CoreModel, lemma: Lemma, targetPos: PartOfSpeech): Sequence<Lex>? {
        return model.lexFinder(lemma)
            ?.asSequence()
            ?.filter { it.partOfSpeech.value == targetPos.value }
    }

    /**
     * Find lexes matching lemma ignoring case and having the desired type
     *
     * @param model      model
     * @param lemma      target lemma (its case is immaterial)
     * @param targetType target typer
     * @return sequence of lexes
     */
    fun getLcLexesHavingType(model: CoreModel, lemma: Lemma, targetType: SynsetType): Sequence<Lex>? {
        return model.lexIgnoreCaseFinder(lemma)
            ?.asSequence()
            ?.filter { it.type.value == targetType.value }
    }

    /**
     * Find lexes matching lemma ignoring case and having the desired pos
     *
     * @param model          model
     * @param lemma          target lemma (its case is immaterial)
     * @param targetPos      target part-of-speech
     * @return sequence of lexes
     */
    fun getLcLexesHavingPos(model: CoreModel, lemma: Lemma, targetPos: PartOfSpeech): Sequence<Lex>? {
        return model.lexIgnoreCaseFinder(lemma)
            ?.asSequence()
            ?.filter { it.partOfSpeech.value == targetPos.value }
    }

    /**
     * Find lexes matching lemma ignoring case
     *
     * @param model         model
     * @param lemma         target lemma (its case is immaterial)
     * @return sequence of lexes
     */
    fun getLcLexes(model: CoreModel, lemma: Lemma): Sequence<Lex>? {
        return model.lexIgnoreCaseFinder(lemma)
            ?.asSequence()
    }

    /**
     * Find lexes matching pronunciations
     *
     * @param lexes          lexes
     * @param pronunciations pronunciations
     * @return sequence of lexes
     * @throws IllegalArgumentException if not found
     */
    fun getLexesHavingPronunciations(lexes: Sequence<Lex>, pronunciations: Set<Pronunciation>?): Sequence<Lex> {
        val p = pronunciations ?: emptySet()
        return lexes
            .filter { p == it.pronunciations }
    }

    /**
     * Find lex matching discriminant
     *
     * @param lexes        lexes
     * @param discriminant discriminant
     * @return sequence of lexes
     * @throws IllegalArgumentException if not found
     */
    fun getLexesHavingDiscriminant(lexes: Sequence<Lex>, discriminant: Discriminant?): Sequence<Lex> {
        return lexes
            .filter { it.discriminant == discriminant }
    }

    // Set comparison

    /**
     * Compare two arrays as sets
     *
     * @param array1 array 1
     * @param array2 array 2
     * @param T type of element
     * @return true if equals
     */
    fun <T> compareAsSets(array1: Array<T>?, array2: Array<T>?): Boolean {
        val set1 = array1?.toSet() ?: emptySet()
        return compareAsSets(set1, array2)
    }

    /**
     * Compare set to array as set
     *
     * @param set1   set
     * @param array2 array
     * @param T type of element
     * @return true if equals
     */
    private fun <T> compareAsSets(set1: Set<T>, array2: Array<T>?): Boolean {
        val set2 = array2?.toSet() ?: emptySet()
        return set1 == set2
    }
}
